libraries/AP_SmartAudio/AP_SmartAudio.cpp

```diff
@@ -0,0 +1,955 @@
+   smartaudioSettings_t _current_state;
+
+    // peek from buffer
+   vtx_states_queue.peek(&_current_state,1);
+    smartaudioFrame_t request;
+    uint8_t frame_size=0;
+
+    debug("%80s::set_power_dbm(%d)\t",TAG,power);
+    frame_size=smartaudioFrameSetPower(&request,_get_power_level_from_dbm(_current_state.version,power));
+     packet command;
+     command.frame=request;
+     command.frame_size=frame_size;
+     requests_queue.push_force(command);
+}
+
+// send vtx request to set power defined in mw
+void AP_SmartAudio::set_power_mw(uint16_t power_mw){
+    set_power_dbm(_get_power_in_dbm_from_mw(power_mw));
+    return;
+}
+
+// returns the power_level applicable when request set power, version 2.1 return the MSB power bit masked at 1
+uint8_t AP_SmartAudio::_get_power_level_from_dbm(uint8_t sma_version,uint8_t power){
+    uint16_t powerLevel=0x00;
+    if (sma_version!=SMARTAUDIO_SPEC_PROTOCOL_v1 && sma_version!=SMARTAUDIO_SPEC_PROTOCOL_v2 && sma_version!=SMARTAUDIO_SPEC_PROTOCOL_v21) {
+        return 0;
+    }
+
+    // version 1
+    if (sma_version==SMARTAUDIO_SPEC_PROTOCOL_v1) {
+        if (power==14) {
+            powerLevel=7;
+        }
+        if (power==23) {
+            powerLevel=16;
+        }
+        if (power==27) {
+            powerLevel=25;
+        }
+        if (power==29) {
+            powerLevel=40;
+        }
+       return powerLevel;
+    }
+    // hardcoded protocol version 2
+    if (sma_version==SMARTAUDIO_SPEC_PROTOCOL_v2) {
+        if (power==14) {
+            powerLevel=0;
+        }
+        if (power==23) {
+            powerLevel=1;
+        }
+        if (power==27) {
+            powerLevel=2;
+        }
+        if (power==29) {
+            powerLevel=3;
+        }
+       return powerLevel;
+    }
+    // hardcoded protocol version 2.1
+    if (sma_version==SMARTAUDIO_SPEC_PROTOCOL_v21) {
+        if (power==14) {
+            powerLevel=14;
+        }
+        if (power==23) {
+            powerLevel=23;
+        }
+        if (power==27) {
+            powerLevel=27;
+        }
+        if (power==29) {
+            powerLevel=29;
+        }
+        return powerLevel|=128;
+    }
+     return 0;
+
+}
+
+/*
+ * This file is part of Cleanflight and Betaflight.
+ *
+ * Cleanflight and Betaflight are free software. You can redistribute
+ * this software and/or modify this software under the terms of the
+ * GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * Cleanflight and Betaflight are distributed in the hope that they
+ * will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
```

Comment:
These need to be converted to AP naming and made part of the class

Reply:
So the smartaudio_protocol.h must be in AP_SmartAudio too, isn't it?

Reply:
Yes I suggest so